GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,208 @@
+        {
+            if (string.IsNullOrEmpty(this.certificatePathOrSubjectCommonName))
+            {
+                return null;
+            }
+
+            EventMetadata metadata = new EventMetadata
+            {
+                { "CertificatePathOrSubjectCommonName", this.certificatePathOrSubjectCommonName },
+                { "IsCertificatePasswordProtected", this.isCertificatePasswordProtected },
+                { "ShouldVerify", this.ShouldVerify }
+            };
+
+            X509Certificate2 result =
+                this.GetCertificateFromFile(tracer, metadata, gitProcess) ??
+                this.GetCertificateFromStore(tracer, metadata);
+
+            if (result == null)
+            {
+                tracer.RelatedError("Certificate {0} not found", this.certificatePathOrSubjectCommonName);
+            }
+
+            return result;
+        }
+
+        private static void LogWithAppropriateLevel(ITracer tracer, EventMetadata metadata, IEnumerable<X509Certificate2> certificates, string logMessage)
+        {
+            Action<EventMetadata, string> loggingFunction;
+            int numberOfCertificates = certificates.Count();
+
+            switch (numberOfCertificates)
+            {
+                case 0:
+                    loggingFunction = tracer.RelatedError;
+                    break;
+                case 1:
+                    loggingFunction = tracer.RelatedInfo;
+                    break;
+                default:
+                    loggingFunction = tracer.RelatedWarning;
+                    break;
+            }
+
+            loggingFunction(
+                metadata,
+                logMessage);
+        }
+
+        private static string GetSubjectNameLineForLogging(IEnumerable<X509Certificate2> certificates)
+        {
+            return string.Join(
+                        Environment.NewLine,
+                        certificates.Select(x => x.Subject));
+        }
+
+        private X509Certificate2 GetCertificateFromFile(ITracer tracer, EventMetadata metadata, GitProcess gitProcess)
+        {
+            string certificatePassword = null;
+            if (this.isCertificatePasswordProtected)
+            {
+                certificatePassword = this.GetCertificatePassword(tracer, gitProcess);
+
+                if (string.IsNullOrEmpty(certificatePassword))
+                {
+                    tracer.RelatedWarning(
+                        new EventMetadata
+                        {
+                                { "SslCertificate", this.certificatePathOrSubjectCommonName }
+                        },
+                        "Git config indicates, that certificate is password protected, but retrieved password was null or empty!");
+                }
+
+                metadata.Add("isPasswordSpecified", string.IsNullOrEmpty(certificatePassword));
+            }
+
+            if (File.Exists(this.certificatePathOrSubjectCommonName))

```

Comment:
For unit testing `PhysicalFileSystem.FileExists` should be used instead of `File.Exists`